### PR TITLE
feat(api): Add async flavor for request interceptors

### PR DESCRIPTION
**Pull request by royjit**
## Description

This is a feature branch PR. All the commits in this PR are already reviewed and approved.
https://github.com/aws-amplify/amplify-swift/pull/2870 feat(api): Add async protocol implementation for token request interceptors
https://github.com/aws-amplify/amplify-swift/pull/2871 feat(api): Async IAM signing
https://github.com/aws-amplify/amplify-swift/pull/2876 feat(api): Update URL interceptor to be async

## General Checklist

- [x] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [x] PR title conforms to conventional commit style
- [ ] New or updated tests include `Given When Then` inline code documentation and are named accordingly `testThing_condition_expectation()`
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/HeaderIAMSigningHelper.swift [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import AWSCore
+import AppSyncRealTimeClient
+
+struct HeaderIAMSigningHelper {
```
Comment (thisisabhash) [inline AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/HeaderIAMSigningHelper.swift:12]: nit: It'd be great to add some documentation about this class. Also, suggestion for naming it IAMHeaderSigningHelper

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift [hunk 2]
```diff
@@ -28,193 +21,137 @@ class IAMAuthInterceptor: AuthInterceptor {
...
-            for (key, value) in headers {
-                try container.encode(value, forKey: DynamicCodingKeys(stringValue: key)!)
+            guard let helper = HeaderIAMSigningHelper(
+                endpoint: endpoint,
+                payload: payload,
+                region: region) else {
+                return nil
             }
+            var signedHeader: IAMAuthenticationHeader?
+            let semaphore = DispatchSemaphore(value: 0)
```
Comment (thisisabhash) [inline AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift:149]: Can this cause threading issues?

**Comment by thisisabhash** [comment]
nit: It'd be great if we can add detailed description about changes in the PR.

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift (hunk 2 above)
Comment (royjit) [reply AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift:149]: Yes it does, this logic keeps the old behavior. But there is another method that adds new async flow without the semaphore logic.

**Review by 5d — APPROVED** [review]

File: AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/HeaderIAMSigningHelper.swift (hunk 1 above)
Comment (royjit) [reply AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/HeaderIAMSigningHelper.swift:12]: Will handle this as a separate PR

**Review by thisisabhash — APPROVED** [review]
LGTM

**Pull request merged**